File: Getting-Started/hello-world/my_agent.py
"""
Copyright (c) 2025 Xpander, Inc. All rights reserved.
"""

import asyncio
import json
import sys
import time
from typing import Optional, List, Dict, Any
from loguru import logger
from xpander_sdk import (
    Agent, LLMProvider, MemoryStrategy, LLMTokens,Tokens
)
from providers.llms.openai.async_client import AsyncOpenAIProvider
from tools.async_function_caller import execute_local_functions
from tools.local_tools import local_tools_by_name, local_tools_list
from dotenv import load_dotenv

load_dotenv()

logger.remove()  # Remove all handlers including the default one
logger.add(sink=sys.stderr, level="INFO")

# Provider to use. Default OpenAI
llm_provider = LLMProvider.OPEN_AI

class MyAgent:
    """
    A framework agnostic agent implementation leveraging xpander.ai's backend-as-a-service infrastructure.

    Handles asynchronous LLM interactions, orchestrates multi-step reasoning, and seamlessly integrates both local and cloud-based tools.

    Args:
        agent (Agent): The xpander.ai Agent instance to operate.
    """

    def __init__(
        self,
        agent: Agent
    ) -> None:
        self.agent = agent
        self.llm_provider = llm_provider

        self.agent.memory_strategy = MemoryStrategy.BUFFERING
        
        with open("agent_instructions.json", "r") as f:
            local_instructions = json.load(f)
        
        if local_instructions:
            self.agent.instructions.role = local_instructions["role"]
            self.agent.instructions.goal = local_instructions["goal"]
            self.agent.instructions.general = local_instructions["general"]
            
        # Register local tools with the agent
        self.agent.add_local_tools(local_tools_list)
        
        self.agent.select_llm_provider(llm_provider)

        if llm_provider == LLMProvider.OPEN_AI:
            self.model_endpoint = AsyncOpenAIProvider()

    async def chat(self, user_input: str, thread_id: Optional[str] = None) -> str:
        """
        Public entry point for chat interaction.

        Adds a user task to the agent memory and initiates the async reasoning loop.

        Args:
            user_input (str): User's input or instruction.
            thread_id (Optional[str]): Memory thread to append to (if continuing a thread).

        Returns:
            str: The memory thread ID of the resulting agent run.
        """
        if thread_id:
            logger.info(f"🧠 Adding task to existing thread: {thread_id}")
            self.agent.add_task(input=user_input, thread_id=thread_id)
        else:
            logger.info("🧠 Adding task to a new thread")
            self.agent.add_task(input=user_input)

        agent_thread = await self._agent_loop()
        logger.info("-" * 80)
        logger.info(f"🤖 Agent response: {agent_thread.result}")
        return agent_thread.memory_thread_id

    async def _call_model(self) -> Dict[str, Any]:
        """
        Internal helper to call the model endpoint.

        Args:
            tools (Optional[List[Dict]]): Tool specification for the model.

        Returns:
            Dict[str, Any]: Model response or error.
        """
        response = await self.model_endpoint.invoke_model(
            messages=self.agent.messages,
            temperature=0.0,
            tools=self.agent.get_tools(),
            tool_choice=self.agent.tool_choice,
        )

        return response

    async def _agent_loop(self):
        """
        Core async loop coordinating reasoning steps and tool execution.

        Returns:
            Any: The final agent thread result after loop completion.
        """
        step = 1
        logger.info("🪄 Starting Agent Loop")
        execution_tokens = Tokens(worker=LLMTokens(0, 0, 0))
        execution_start_time = time.perf_counter()

        ## main agentic loop
        while not self.agent.is_finished():
            logger.info("-" * 80)
            logger.info(f"🔍 Step {step}")

            ## call the AI model with tools and the state of the agent
            response = await self._call_model()

            ## handle token accounting
            step_usage = self.model_endpoint.handle_token_accounting(
                execution_tokens=execution_tokens,
                response=response,
            )

            ## add llm response to agent memory
            llm_response = response.model_dump() if not isinstance(response, dict) else response
            self.agent.add_messages(llm_response)
            
            ## report execution metrics
            self.agent.report_execution_metrics(
                llm_tokens=execution_tokens,
                ai_model=self.model_endpoint.model_id,
            )

            ## extract all tool calls from llm response
            tool_calls = self.agent.extract_tool_calls(llm_response=llm_response)

            ## execute cloud tool calls
            cloud_tool_call_results = await asyncio.to_thread(
                self.agent.run_tools,
                tool_calls=tool_calls,
            )

            ## retrieve pending local tool calls
            local_tool_calls = await asyncio.to_thread(
                self.agent.retrieve_pending_local_tool_calls,
                tool_calls=tool_calls,
            )
            
            ## remove cloud tool calls empty results that are already in local tool calls
            cloud_tool_call_results[:] = [
                c for c in cloud_tool_call_results
                if c.tool_call_id not in {t.tool_call_id for t in local_tool_calls}
            ]
            
            ## execute local tool calls
            local_tool_call_results = await execute_local_functions(
                function_calls=local_tool_calls,
                functions_by_name=local_tools_by_name,
            )

            ## add local tool call results to agent memory
            if local_tool_call_results:
                self.agent.memory.add_tool_call_results(
                    tool_call_results=local_tool_call_results,
                )

            ## log tool call results
            for res in cloud_tool_call_results + local_tool_call_results:
                emoji = "✅" if res.is_success else "❌"
                logger.info(f"{emoji} {res.function_name}")

            logger.info(
                f"🔢 Step {step} tokens used: {step_usage.total_tokens} "
                f"(output: {step_usage.completion_tokens}, input: {step_usage.prompt_tokens})"
            )
            step += 1

        logger.info(
            f"✨ Execution duration: {time.perf_counter() - execution_start_time:.2f} s"
        )
        logger.info(
            f"🔢 Total tokens used: {execution_tokens.worker.total_tokens} "
            f"(output: {execution_tokens.worker.completion_tokens}, "
            f"input: {execution_tokens.worker.prompt_tokens})"
        )

        ## return the final agent thread result
        return self.agent.retrieve_execution_result()

File: Getting-Started/node/hello-world/MyAgent.js
/**
 * Copyright (c) 2025 Xpander, Inc. All rights reserved.
 */

import { XpanderClient, LLMProvider, MemoryStrategy , Agent} from 'xpander-sdk';
import OpenAI from 'openai';
import fs from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Provider to use. Default OpenAI
const llmProvider = LLMProvider.OPEN_AI;

export class MyAgent {
    /**
     * A framework agnostic agent implementation leveraging xpander.ai's backend-as-a-service infrastructure.
     * 
     * Handles asynchronous LLM interactions, orchestrates multi-step reasoning, and seamlessly integrates both local and cloud-based tools.
     * 
     * @param {Agent} agent - The xpander.ai Agent instance to operate.
     */
    constructor(agent) {
        this.agent = agent;
        this.llmProvider = llmProvider;
        
        // Set memory strategy using camelCase
        this.agent.memoryStrategy = MemoryStrategy.BUFFERING;
        
        // Load local instructions
        const localInstructions = JSON.parse(
            fs.readFileSync(join(__dirname, 'agent_instructions.json'), 'utf8')
        );
        
        if (localInstructions) {
            this.agent.instructions.role = localInstructions.role;
            this.agent.instructions.goal = localInstructions.goal;
            this.agent.instructions.general = localInstructions.general;
        }
        
        this.agent.selectLLMProvider(llmProvider);
        
        if (llmProvider === LLMProvider.OPEN_AI) {
            this.openaiClient = new OpenAI({
                apiKey: process.env.OPENAI_API_KEY
            });
        }
    }

    /**
     * Public entry point for chat interaction.
     * 
     * Adds a user task to the agent memory and initiates the async reasoning loop.
     * 
     * @param {string} userInput - User's input or instruction.
     * @param {string} [threadId] - Memory thread to append to (if continuing a thread).
     * @returns {Promise<string>} The memory thread ID of the resulting agent run.
     */
    async chat(userInput, threadId = null) {
        if (threadId) {
            console.log(`🧠 Adding task to existing thread: ${threadId}`);
            // Try using object pattern to skip middle parameters
            this.agent.addTask(userInput, threadId);
        } else {
            console.log("🧠 Adding task to a new thread");
            this.agent.addTask(userInput);
        }

        const agentThread = await this._agentLoop();
        console.log("-".repeat(80));
        console.log(`🤖 Agent response: ${agentThread.result}`);
        return agentThread.memoryThreadId;
    }

    /**
     * Internal helper to call the model endpoint.
     * 
     * @returns {Promise<Object>} Model response or error.
     */
    async _callModel() {
        const response = await this.openaiClient.chat.completions.create({
            messages: this.agent.messages,
            temperature: 0.0,
            tools: this.agent.getTools(),
            tool_choice: this.agent.toolChoice,
            model: "gpt-4o"
        });

        return response;
    }

    /**
     * Core async loop coordinating reasoning steps and tool execution.
     * 
     * @returns {Promise<any>} The final agent thread result after loop completion.
     */
    async _agentLoop() {
        let step = 1;
        console.log("🪄 Starting Agent Loop");
        const executionTokens = { worker: { totalTokens: 0, completionTokens: 0, promptTokens: 0 } };
        const executionStartTime = Date.now();

        // Main agentic loop
        while (!this.agent.isFinished()) {
            console.log("-".repeat(80));
            console.log(`🔍 Step ${step}`);

            // Call the AI model with tools and the state of the agent
            const response = await this._callModel();

            // Handle token accounting
            const stepUsage = this._handleTokenAccounting(executionTokens, response);

            // Add llm response to agent memory using camelCase
            const llmResponse = response.toJSON ? response.toJSON() : response;
            this.agent.addMessages(llmResponse);
            
            // Report execution metrics using camelCase
            this.agent.reportExecutionMetrics(
                executionTokens,
                this.openaiClient.model || "gpt-4o"
            );

            // Extract all tool calls from llm response using camelCase
            const toolCalls = XpanderClient.extractToolCalls(llmResponse);

            // Execute cloud tool calls using camelCase
            const cloudToolCallResults = await this.agent.runTools(toolCalls);

            // Log tool call results
            for (const res of cloudToolCallResults) {
                const emoji = res.isSuccess ? "✅" : "❌";
                console.log(`${emoji} ${res.functionName}`);
            }

            console.log(
                `🔢 Step ${step} tokens used: ${stepUsage.total_tokens} ` +
                `(output: ${stepUsage.completion_tokens}, input: ${stepUsage.prompt_tokens})`
            );
            step++;
        }

        console.log(`✨ Execution duration: ${(Date.now() - executionStartTime) / 1000}s`);
        console.log(
            `🔢 Total tokens used: ${executionTokens.worker.totalTokens} ` +
            `(output: ${executionTokens.worker.completionTokens}, ` +
            `input: ${executionTokens.worker.promptTokens})`
        );

        // Return the final agent thread result using camelCase
        return this.agent.retrieveExecutionResult();
    }

    /**
     * Handle token accounting for the response
     * 
     * @param {Object} executionTokens - Token tracking object
     * @param {Object} response - OpenAI response
     * @returns {Object} Step usage information
     */
    _handleTokenAccounting(executionTokens, response) {
        const usage = response.usage || { 
            total_tokens: 0, 
            completion_tokens: 0, 
            prompt_tokens: 0 
        };
        
        executionTokens.worker.totalTokens += usage.total_tokens;
        executionTokens.worker.completionTokens += usage.completion_tokens;
        executionTokens.worker.promptTokens += usage.prompt_tokens;
        
        return usage;
    }
} 